addons/website_slides/static/src/js/slides_course_page.js

```diff
@@ -0,0 +1,148 @@
+
+        const $completed = $('.o_wslides_channel_completion_completed');
+        const $progressbar = $('.o_wslides_channel_completion_progressbar');
+
+        if (completion < 100) {
+            // Hide the "Completed" text and show the progress bar
+            $completed.addClass('d-none');
+            $progressbar.removeClass('d-none').addClass('d-flex');
+        } else {
+            // Hide the progress bar and show the "Completed" text
+            $completed.removeClass('d-none');
+            $progressbar.addClass('d-none').removeClass('d-flex');
+        }
+
+        $progressbar.find('.progress-bar').css('width', `${completion}%`);
+        $progressbar.find('.o_wslides_progress_percentage').text(completion);
+    },
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Once the completion conditions are filled,
+     * rpc call to set the relation between the slide and the user as "completed"
+     *
+     * @private
+     * @param {Integer} slideId: the id of slide to set as completed
+     * @param {Boolean} completed: true to mark the slide as completed
+     *     false to mark the slide as not completed
+     */
+    _toggleSlideCompleted: async function (slide, completed = true) {
+        if (!!slide.completed === !!completed || !slide.isMember) {
+            // no useless RPC call
+            return;
+        }
+
+        const data = await this._rpc({
+            route: `/slides/slide/${completed ? 'set_completed' : 'set_uncompleted'}`,
+            params: {slide_id: slide.id},
+        });
+
+        this.toggleCompletionButton(slide, completed);
+        this.updateProgressbar(data.channel_completion);
+
+        slide.completed = true;
+    },
+
+    _getSlide: function (slideId) {
+        return $(`.o_wslides_sidebar_done_button[data-id="${slideId}"]`).data();
+    },
+
+    //--------------------------------------------------------------------------
+    // Handler
+    //--------------------------------------------------------------------------
+    _onClickComplete: function (ev) {
+        ev.stopPropagation();
+        ev.preventDefault();
+
+        const $button = $(ev.currentTarget).parents('.o_wslides_sidebar_done_button');
+
+        const slideData = $button.data();
+        const isCompleted = Boolean(slideData.completed);
+
+        this._toggleSlideCompleted(slideData, !isCompleted);
+    },
+
+    /**
+     * The slide has been completed, update the UI
+     */
+    _onSlideCompleted: function (ev) {
+        const slideId = ev.data.slideId;
+        const completed = ev.data.completed;
+        const slide = this._getSlide(slideId);
+        if (slide) {
+            // Just joined the course (e.g. When "Submit & Join" action), update the UI
+            this.toggleCompletionButton(slide, completed);
```

Comment:
Maybe my "completion widget" idea will not work because of this, as the slide also needs to toggle the button manually.

Sad reviewer is sad.